File: examples/dekita/.claude/hooks/session-worktree-status.py
#!/usr/bin/env python3
"""セッション開始時に既存worktreeの状況を確認し警告する。

Why:
    別セッションが作業中のworktreeに介入すると、競合やコンフリクトが発生する。
    セッション開始時に既存worktreeの状況を把握することで、問題を未然に防ぐ。

What:
    - CWDがworktree内かどうかをチェック
    - 既存worktreeの一覧を取得（git worktree list）
    - 各worktreeの.claude-sessionマーカーを確認
    - 別セッションIDのマーカーがある場合、警告表示
    - 直近1時間以内のコミットがあるworktreeも警告
    - fork-sessionの場合、祖先セッションのworktreeへの介入を禁止警告

State:
    reads: .worktrees/*/.claude-session
    reads: .claude/logs/flow/session-created-issues-*.json

Remarks:
    - ブロックせず警告のみ（実際のブロックはworktree-session-guard.pyが担当）
    - fork-session-collaboration-advisorは提案、これは警告

Changelog:
    - silenvx/dekita#1383: CWDがworktree内かどうかの検出追加
    - silenvx/dekita#1416: フック追加
    - silenvx/dekita#2466: fork-sessionの祖先worktree介入警告追加
    - silenvx/dekita#2475: fork-sessionで自作Issue許可
"""

import json
import os
import shlex
import subprocess
import sys
import time
from datetime import UTC, datetime
from pathlib import Path
from typing import TypedDict

from lib.constants import (
    RECENT_COMMIT_THRESHOLD_SECONDS,
    SESSION_GAP_THRESHOLD,
    SESSION_MARKER_FILE,
    TIMEOUT_LIGHT,
    TIMEOUT_MEDIUM,
)
from lib.execution import log_hook_execution
from lib.session import (
    create_hook_context,
    get_session_ancestry,
    is_fork_session,
    parse_hook_input,
)


def _get_session_log_dir() -> Path:
    """Get the directory for session-specific log files.

    Uses .claude/logs/flow/ for consistency with other session logs.
    """
    project_dir = os.environ.get("CLAUDE_PROJECT_DIR", os.getcwd())
    return Path(project_dir) / ".claude" / "logs" / "flow"


def load_session_created_issues(session_id: str) -> list[int]:
    """Load list of issue numbers created in this session.

    Issue #2475: Used to identify self-created Issues in fork-sessions.

    Args:
        session_id: The Claude session ID.

    Returns:
        List of issue numbers created in this session.
    """
    # Sanitize session_id to prevent path traversal (e.g., "../../etc/passwd")
    safe_session_id = Path(session_id).name
    issues_file = _get_session_log_dir() / f"session-created-issues-{safe_session_id}.json"
    if issues_file.exists():
        try:
            data = json.loads(issues_file.read_text())
            # Guard against non-dict JSON (e.g., array or string)
            if isinstance(data, dict):
                issues = data.get("issues", [])
                # Type validation: must be list of ints
                if isinstance(issues, list) and all(isinstance(n, int) for n in issues):
                    return issues
        except (json.JSONDecodeError, OSError):
            pass  # Best effort - corrupted data is ignored
    return []


# Worktrees directory name
WORKTREES_DIR = ".worktrees"


def get_cwd_worktree_info() -> tuple[str, Path] | None:
    """Check if current working directory is inside a worktree.

    Returns:
        Tuple of (worktree_name, main_repo_path) if CWD is in a worktree,
        None otherwise.
    """
    try:
        cwd = Path.cwd()
    except OSError:
        # CWD access error (deleted directory, etc.)
        return None

    # Find .worktrees in path
    parts = cwd.parts
    for i, part in enumerate(parts):
        if part == WORKTREES_DIR:
            # Next part after .worktrees is the worktree name
            if i + 1 < len(parts):
                worktree_name = parts[i + 1]
                # Main repo path is everything before .worktrees
                main_repo_path = Path(*parts[:i])
                return (worktree_name, main_repo_path)
    return None


class WorktreeInfo(TypedDict):
    """Information about a worktree."""

    path: Path
    locked: bool


def get_worktrees_info() -> list[WorktreeInfo]:
    """Get list of worktree directories and their lock status from git.

    This function makes a single git call to get both worktree paths and
    lock status, avoiding redundant subprocess calls.

    Returns:
        List of WorktreeInfo dictionaries containing 'path' (Path) and 'locked' (bool).
    """
    try:
        result = subprocess.run(
            ["git", "worktree", "list", "--porcelain"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
        )
        if result.returncode != 0:
            return []

        worktrees_info: list[WorktreeInfo] = []
        # Split by double newline to get blocks for each worktree
        blocks = result.stdout.strip().split("\n\n")
        for block in blocks:
            lines = block.splitlines()
            if not lines or not lines[0].startswith("worktree "):
                continue

            path = Path(lines[0][9:].strip())

            # Only include worktrees in .worktrees directory
            if WORKTREES_DIR not in path.parts:
                continue

            # Check if locked (line starts with "locked" or "locked <reason>")
            is_locked = any(line.startswith("locked") for line in lines)
            worktrees_info.append({"path": path, "locked": is_locked})

        return worktrees_info
    except (subprocess.TimeoutExpired, OSError):
        # Git command failures are treated as "no worktrees"
        # to fail-open and not block session start unnecessarily
        return []


class SessionMarker(TypedDict, total=False):
    """Session marker data from worktree marker file."""

    session_id: str
    created_at: str  # ISO format timestamp


def read_session_marker(worktree_path: Path) -> SessionMarker | None:
    """Read session marker from worktree marker file.

    Expects JSON format:
    {
        "session_id": "...",
        "created_at": "2025-12-30T09:30:00+00:00"
    }

    Args:
        worktree_path: Path to worktree directory

    Returns:
        SessionMarker dict if marker exists and is valid JSON, None otherwise.
    """
    marker_path = worktree_path / SESSION_MARKER_FILE
    try:
        if marker_path.exists():
            content = marker_path.read_text().strip()
            data = json.loads(content)
            return {
                "session_id": data.get("session_id", ""),
                "created_at": data.get("created_at", ""),
            }
    except (OSError, json.JSONDecodeError):
        # File access errors or invalid JSON are treated as "no marker"
        # to fail-open and not block session start unnecessarily
        pass
    return None


def get_marker_age_seconds(marker: SessionMarker) -> int | None:
    """Get age of marker in seconds from created_at timestamp.

    Args:
        marker: Session marker dict

    Returns:
        Age in seconds if created_at exists and is valid, None otherwise.
    """
    created_at = marker.get("created_at", "")
    if not created_at:
        return None
    try:
        created_time = datetime.fromisoformat(created_at)
        now = datetime.now(UTC)
        # Ensure both are timezone-aware for comparison
        if created_time.tzinfo is None:
            created_time = created_time.replace(tzinfo=UTC)
        return int((now - created_time).total_seconds())
    except (ValueError, TypeError):
        return None


def get_recent_commit_time(worktree_path: Path) -> int | None:
    """Get seconds since the most recent commit in the worktree.

    Args:
        worktree_path: Path to worktree directory

    Returns:
        Seconds since last commit, or None if unable to determine.
    """
    try:
        result = subprocess.run(
            ["git", "-C", str(worktree_path), "log", "-1", "--format=%ct"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_LIGHT,
        )
        if result.returncode == 0 and result.stdout.strip():
            commit_time = int(result.stdout.strip())
            return int(time.time()) - commit_time
    except (subprocess.TimeoutExpired, OSError, ValueError):
        # Git command failures or timeouts are treated as "unable to determine"
        # to fail-open and not block session start unnecessarily
        pass
    return None


def has_uncommitted_changes(worktree_path: Path) -> bool:
    """Check if worktree has uncommitted changes.

    Note: This function is named differently from lib/git.check_uncommitted_changes
    to avoid name collision. This version uses fail-open policy and returns bool only,
    while lib/git version uses fail-close and returns tuple[bool, int].

    Args:
        worktree_path: Path to worktree directory

    Returns:
        True if there are uncommitted changes, False otherwise (including on error).
    """
    try:
        result = subprocess.run(
            ["git", "-C", str(worktree_path), "status", "--porcelain"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_LIGHT,
        )
        return result.returncode == 0 and bool(result.stdout.strip())
    except (subprocess.TimeoutExpired, OSError):
        # Fail-open: treat errors as "no uncommitted changes"
        return False


def main():
    """SessionStart hook to warn about existing worktrees."""
    try:
        # Parse input and set session ID
        input_data = parse_hook_input()
        ctx = create_hook_context(input_data)
        current_session = ctx.get_session_id()

        # Issue #2466: Detect fork-session and get ancestor sessions
        source = input_data.get("source", "")
        transcript_path = input_data.get("transcript_path")
        is_fork = is_fork_session(current_session, source, transcript_path)
        ancestor_sessions: list[str] = []
        if is_fork and transcript_path:
            ancestor_sessions = get_session_ancestry(transcript_path)
            # Remove current session from ancestors if present
            ancestor_sessions = [s for s in ancestor_sessions if s != current_session]

        warnings: list[str] = []
        fork_session_warnings: list[str] = []  # Issue #2466: Separate list for fork warnings
        cwd_warning_prefix = ""

        # Check if CWD is inside a worktree (Issue #1383)
        cwd_worktree_info = get_cwd_worktree_info()
        if cwd_worktree_info:
            cwd_worktree_name, main_repo_path = cwd_worktree_info
            # Use shlex.quote to properly escape path for shell (handles quotes, spaces, etc.)
            quoted_path = shlex.quote(str(main_repo_path))
            cwd_warning_prefix = (
                f"**CWDがworktree内です: {cwd_worktree_name}**\n\n"
                "セッション継続後もCWDがworktree内のままになっています。\n"
                "worktree削除がブロックされる可能性があるため、mainリポジトリに移動してください:\n\n"
                f"```\ncd {quoted_path}\n```\n\n"
            )
            log_hook_execution(
                "session-worktree-status", "warn", f"CWD is inside worktree: {cwd_worktree_name}"
            )

        # Get list of worktrees with lock status (single git call)
        worktrees_info = get_worktrees_info()
        if not worktrees_info and not cwd_warning_prefix:
            log_hook_execution("session-worktree-status", "skip", "No worktrees found")
            print(json.dumps({"continue": True}))
            return

        # Check each worktree for issues
        for info in worktrees_info:
            worktree_path = info["path"]
            worktree_name = worktree_path.name
            issues: list[str] = []

            # Check session marker
            marker = read_session_marker(worktree_path)
            if marker:
                marker_session = marker.get("session_id", "")
                marker_age = get_marker_age_seconds(marker)

                if marker_session and marker_session != current_session:
                    # Different session ID
                    session_display = (
                        f"{marker_session[:16]}..." if len(marker_session) > 16 else marker_session
                    )

                    # Issue #2466: Check if this is an ancestor session's worktree
                    if is_fork and marker_session in ancestor_sessions:
                        # This is a fork-session trying to access parent session's worktree
                        fork_session_warnings.append(
                            f"- **{worktree_name}**: 元セッション（fork元）のworktree"
                        )
                        log_hook_execution(
                            "session-worktree-status",
                            "warn",
                            f"Fork-session detected ancestor worktree: {worktree_name}",
                        )
                    else:
                        issues.append(f"別セッション: {session_display}")
                elif marker_session == current_session and marker_age is not None:
                    # Same session ID but check if marker is stale (context continuation case)
                    # If marker is older than SESSION_GAP_THRESHOLD, warn about potential
                    # context continuation from a different "instance" of the same session
                    if marker_age > SESSION_GAP_THRESHOLD:
                        minutes = marker_age // 60
                        issues.append(f"古いセッションマーカー: {minutes}分前")

            # Check if locked (already retrieved in get_worktrees_info)
            if info["locked"]:
                issues.append("ロック中")

            # Check uncommitted changes
            if has_uncommitted_changes(worktree_path):
                issues.append("未コミット変更あり")

            # Check recent commits
            seconds_since_commit = get_recent_commit_time(worktree_path)
            if seconds_since_commit is not None:
                if seconds_since_commit < RECENT_COMMIT_THRESHOLD_SECONDS:
                    minutes = seconds_since_commit // 60
                    # Use "1分未満" for commits less than 1 minute ago
                    if minutes < 1:
                        issues.append("直近1分未満にコミット")
                    else:
                        issues.append(f"直近{minutes}分前にコミット")

            if issues:
                warnings.append(f"- **{worktree_name}**: {', '.join(issues)}")

        # Build final warning message
        if warnings or cwd_warning_prefix or fork_session_warnings:
            warning_parts = ["[session-worktree-status]"]

            # Add CWD warning if present
            if cwd_warning_prefix:
                warning_parts.append(cwd_warning_prefix)

            # Issue #2466: Add fork-session warning (after CWD warning if present)
            # Issue #2470: 自分で作成したIssueは例外
            # Issue #2475: 自作Issueリストを表示
            if fork_session_warnings:
                # Load self-created Issues for this session
                self_created_issues = load_session_created_issues(current_session)
                if self_created_issues:
                    issue_list = ", ".join(f"#{n}" for n in self_created_issues)
                    self_created_note = (
                        f"\n**このセッションで作成したIssue**: {issue_list}\n"
                        "これらのIssueへの作業は**警告なしで許可**されています。\n"
                    )
                else:
                    self_created_note = (
                        "\n**このセッションで作成したIssue**: "
                        "（このセッション中に作成したIssueはまだありません）\n"
                        "今後このセッションで新しく作成したIssueへの作業は"
                        "**警告なしで許可**されます。\n"
                    )

                warning_parts.append(
                    "\n".join(
                        [
                            "⚠️ **fork-session検出**: 元セッション（fork元）のworktreeがあります:",
                            "",
                            *fork_session_warnings,
                            "",
                            "**これらのworktreeへの介入は禁止です。**",
                            "元セッションがまだ作業中の可能性があります。",
                            self_created_note,
                            "その場合は、元セッションのworktreeとは**別の新しいworktree**を作成してください。",
                            "",
                        ]
                    )
                )

            # Add worktree warnings if present
            if warnings:
                warning_parts.append(
                    "以下のworktreeに注意が必要です:\n\n" + "\n".join(warnings) + "\n\n"
                    "これらのworktreeに関連するIssueは作業中の可能性があります。\n"
                    "AGENTS.mdのルールに従い、以下を確認してください:\n"
                    "- worktreeがロック中 or 未コミット変更あり → 作業開始しない\n"
                    "- 直近1時間以内のコミット → ユーザーに確認\n"
                    "- 別セッションマーカー → 引き継がず他のIssueへ\n"
                    "- 古いセッションマーカー → コンテキスト継続の可能性。ユーザーに確認\n"
                )

            warning_msg = " ".join(warning_parts[:2])
            if len(warning_parts) > 2:
                warning_msg += "\n\n" + "\n\n".join(warning_parts[2:])

            log_hook_execution(
                "session-worktree-status",
                "warn",
                f"Found {len(warnings)} worktrees requiring attention, "
                f"fork-session warnings: {len(fork_session_warnings)}, "
                f"CWD in worktree: {bool(cwd_warning_prefix)}",
            )
            print(json.dumps({"continue": True, "systemMessage": warning_msg}))
        else:
            log_hook_execution(
                "session-worktree-status",
                "approve",
                f"Checked {len(worktrees_info)} worktrees, all clear",
            )
            print(json.dumps({"continue": True}))

    except Exception as e:
        # Fail open - don't block on errors
        error_msg = f"Hook error: {e}"
        print(f"[session-worktree-status] {error_msg}", file=sys.stderr)
        print(json.dumps({"continue": True}))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/session_worktree_status.ts
#!/usr/bin/env bun
/**
 * セッション開始時に既存worktreeの状況を確認し警告する。
 *
 * Why:
 *   別セッションが作業中のworktreeに介入すると、競合やコンフリクトが発生する。
 *   セッション開始時に既存worktreeの状況を把握することで、問題を未然に防ぐ。
 *
 * What:
 *   - CWDがworktree内かどうかをチェック
 *   - 既存worktreeの一覧を取得（git worktree list）
 *   - 各worktreeの.claude-sessionマーカーを確認
 *   - 別セッションIDのマーカーがある場合、警告表示
 *   - 直近1時間以内のコミットがあるworktreeも警告
 *   - fork-sessionの場合、祖先セッションのworktreeへの介入を禁止警告
 *
 * State:
 *   reads: .worktrees/{name}/.claude-session
 *   reads: .claude/logs/flow/session-created-issues-{session}.json
 *
 * Remarks:
 *   - ブロックせず警告のみ（実際のブロックはworktree-session-guard.tsが担当）
 *   - fork-session-collaboration-advisorは提案、これは警告
 *   - Python版: session_worktree_status.py
 *
 * Changelog:
 *   - silenvx/dekita#1383: CWDがworktree内かどうかの検出追加
 *   - silenvx/dekita#1416: フック追加
 *   - silenvx/dekita#2466: fork-sessionの祖先worktree介入警告追加
 *   - silenvx/dekita#2475: fork-sessionで自作Issue許可
 *   - silenvx/dekita#2917: TypeScript版初期実装
 */

import { execSync } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import { basename, join } from "node:path";
import { FLOW_LOG_DIR } from "../lib/common";
import {
  RECENT_COMMIT_THRESHOLD_SECONDS,
  SESSION_GAP_THRESHOLD,
  SESSION_MARKER_FILE,
  TIMEOUT_LIGHT,
  TIMEOUT_MEDIUM,
} from "../lib/constants";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import {
  createContext,
  getSessionAncestry,
  getSessionId,
  isForkSession,
  parseHookInput,
} from "../lib/session";

const HOOK_NAME = "session-worktree-status";
const WORKTREES_DIR = ".worktrees";

interface WorktreeInfo {
  path: string;
  locked: boolean;
}

interface SessionMarker {
  session_id: string;
  created_at: string;
}

/**
 * Sanitize session ID to prevent path traversal.
 */
function sanitizeSessionId(sessionId: string): string {
  return basename(sessionId);
}

/**
 * Load list of issue numbers created in this session.
 */
function loadSessionCreatedIssues(sessionId: string): number[] {
  const safeSessionId = sanitizeSessionId(sessionId);
  const issuesFile = join(FLOW_LOG_DIR, `session-created-issues-${safeSessionId}.json`);

  if (existsSync(issuesFile)) {
    try {
      const content = readFileSync(issuesFile, "utf-8");
      const data = JSON.parse(content);
      if (typeof data === "object" && data !== null && !Array.isArray(data)) {
        const issues = data.issues;
        if (Array.isArray(issues) && issues.every((n: unknown) => typeof n === "number")) {
          return issues;
        }
      }
    } catch {
      // Best effort - corrupted data is ignored
    }
  }
  return [];
}

/**
 * Check if current working directory is inside a worktree.
 */
function getCwdWorktreeInfo(): { worktreeName: string; mainRepoPath: string } | null {
  try {
    const cwd = process.cwd();
    const parts = cwd.split("/");

    for (let i = 0; i < parts.length; i++) {
      if (parts[i] === WORKTREES_DIR) {
        // Next part after .worktrees is the worktree name
        if (i + 1 < parts.length) {
          const worktreeName = parts[i + 1];
          // Main repo path is everything before .worktrees
          const mainRepoPath = parts.slice(0, i).join("/") || "/";
          return { worktreeName, mainRepoPath };
        }
      }
    }
  } catch {
    // CWD access error (deleted directory, etc.)
  }
  return null;
}

/**
 * Get list of worktree directories and their lock status from git.
 */
function getWorktreesInfo(): WorktreeInfo[] {
  try {
    const result = execSync("git worktree list --porcelain", {
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM * 1000,
      stdio: ["pipe", "pipe", "pipe"],
    });

    const worktreesInfo: WorktreeInfo[] = [];
    // Split by double newline to get blocks for each worktree
    const blocks = result.trim().split("\n\n");

    for (const block of blocks) {
      const lines = block.split("\n");
      if (!lines[0]?.startsWith("worktree ")) {
        continue;
      }

      const path = lines[0].slice(9).trim();

      // Only include worktrees in .worktrees directory
      if (!path.includes(`/${WORKTREES_DIR}/`)) {
        continue;
      }

      // Check if locked
      const isLocked = lines.some((line) => line.startsWith("locked"));
      worktreesInfo.push({ path, locked: isLocked });
    }

    return worktreesInfo;
  } catch {
    // Fail-open
    return [];
  }
}

/**
 * Read session marker from worktree marker file.
 */
function readSessionMarker(worktreePath: string): SessionMarker | null {
  const markerPath = join(worktreePath, SESSION_MARKER_FILE);
  try {
    if (existsSync(markerPath)) {
      const content = readFileSync(markerPath, "utf-8").trim();
      const data = JSON.parse(content);
      return {
        session_id: data.session_id || "",
        created_at: data.created_at || "",
      };
    }
  } catch {
    // Fail-open
  }
  return null;
}

/**
 * Get age of marker in seconds from created_at timestamp.
 */
function getMarkerAgeSeconds(marker: SessionMarker): number | null {
  const createdAt = marker.created_at;
  if (!createdAt) {
    return null;
  }
  try {
    const createdTime = new Date(createdAt).getTime();
    const now = Date.now();
    return Math.floor((now - createdTime) / 1000);
  } catch {
    return null;
  }
}

/**
 * Get seconds since the most recent commit in the worktree.
 */
function getRecentCommitTime(worktreePath: string): number | null {
  try {
    const result = execSync(`git -C "${worktreePath}" log -1 --format=%ct`, {
      encoding: "utf-8",
      timeout: TIMEOUT_LIGHT * 1000,
      stdio: ["pipe", "pipe", "pipe"],
    });

    const trimmed = result.trim();
    if (trimmed) {
      const commitTime = Number.parseInt(trimmed, 10);
      return Math.floor(Date.now() / 1000) - commitTime;
    }
  } catch {
    // Fail-open
  }
  return null;
}

/**
 * Check if worktree has uncommitted changes.
 */
function hasUncommittedChanges(worktreePath: string): boolean {
  try {
    const result = execSync(`git -C "${worktreePath}" status --porcelain`, {
      encoding: "utf-8",
      timeout: TIMEOUT_LIGHT * 1000,
      stdio: ["pipe", "pipe", "pipe"],
    });
    return result.trim().length > 0;
  } catch {
    // Fail-open
    return false;
  }
}

/**
 * Escape a path for shell command in markdown code block.
 */
export function shellQuote(path: string): string {
  // If the path contains single quote, use double quotes, otherwise use single quotes
  if (path.includes("'")) {
    return `"${path.replace(/"/g, '\\"')}"`;
  }
  return `'${path}'`;
}

async function main(): Promise<void> {
  const result: { continue: boolean; systemMessage?: string } = { continue: true };

  try {
    const data = await parseHookInput();
    const ctx = createContext(data);
    const currentSession = getSessionId(ctx) || "";

    // Issue #2466: Detect fork-session and get ancestor sessions
    const source = data.source || "";
    const transcriptPath = data.transcript_path;
    const isFork = currentSession ? isForkSession(currentSession, source, transcriptPath) : false;
    let ancestorSessions: string[] = [];
    if (isFork && transcriptPath) {
      ancestorSessions = getSessionAncestry(transcriptPath);
      // Remove current session from ancestors if present
      ancestorSessions = ancestorSessions.filter((s) => s !== currentSession);
    }

    const warnings: string[] = [];
    const forkSessionWarnings: string[] = [];
    let cwdWarningPrefix = "";

    // Check if CWD is inside a worktree (Issue #1383)
    const cwdWorktreeInfo = getCwdWorktreeInfo();
    if (cwdWorktreeInfo) {
      const { worktreeName, mainRepoPath } = cwdWorktreeInfo;
      const quotedPath = shellQuote(mainRepoPath);
      cwdWarningPrefix = `**CWDがworktree内です: ${worktreeName}**\n\nセッション継続後もCWDがworktree内のままになっています。\nworktree削除がブロックされる可能性があるため、mainリポジトリに移動してください:\n\n\`\`\`\ncd ${quotedPath}\n\`\`\`\n\n`;
      await logHookExecution(HOOK_NAME, "warn", `CWD is inside worktree: ${worktreeName}`);
    }

    // Get list of worktrees with lock status (single git call)
    const worktreesInfo = getWorktreesInfo();
    if (worktreesInfo.length === 0 && !cwdWarningPrefix) {
      await logHookExecution(HOOK_NAME, "skip", "No worktrees found");
      console.log(JSON.stringify({ continue: true }));
      return;
    }

    // Check each worktree for issues
    for (const info of worktreesInfo) {
      const worktreePath = info.path;
      const worktreeName = basename(worktreePath);
      const issues: string[] = [];

      // Check session marker
      const marker = readSessionMarker(worktreePath);
      if (marker) {
        const markerSession = marker.session_id;
        const markerAge = getMarkerAgeSeconds(marker);

        if (markerSession && markerSession !== currentSession) {
          // Different session ID
          const sessionDisplay =
            markerSession.length > 16 ? `${markerSession.slice(0, 16)}...` : markerSession;

          // Issue #2466: Check if this is an ancestor session's worktree
          if (isFork && ancestorSessions.includes(markerSession)) {
            forkSessionWarnings.push(`- **${worktreeName}**: 元セッション（fork元）のworktree`);
            await logHookExecution(
              HOOK_NAME,
              "warn",
              `Fork-session detected ancestor worktree: ${worktreeName}`,
            );
          } else {
            issues.push(`別セッション: ${sessionDisplay}`);
          }
        } else if (markerSession === currentSession && markerAge !== null) {
          // Same session ID but check if marker is stale
          if (markerAge > SESSION_GAP_THRESHOLD) {
            const minutes = Math.floor(markerAge / 60);
            issues.push(`古いセッションマーカー: ${minutes}分前`);
          }
        }
      }

      // Check if locked
      if (info.locked) {
        issues.push("ロック中");
      }

      // Check uncommitted changes
      if (hasUncommittedChanges(worktreePath)) {
        issues.push("未コミット変更あり");
      }

      // Check recent commits
      const secondsSinceCommit = getRecentCommitTime(worktreePath);
      if (secondsSinceCommit !== null) {
        if (secondsSinceCommit < RECENT_COMMIT_THRESHOLD_SECONDS) {
          const minutes = Math.floor(secondsSinceCommit / 60);
          if (minutes < 1) {
            issues.push("直近1分未満にコミット");
          } else {
            issues.push(`直近${minutes}分前にコミット`);
          }
        }
      }

      if (issues.length > 0) {
        warnings.push(`- **${worktreeName}**: ${issues.join(", ")}`);
      }
    }

    // Build final warning message
    if (warnings.length > 0 || cwdWarningPrefix || forkSessionWarnings.length > 0) {
      const warningParts: string[] = ["[session-worktree-status]"];

      // Add CWD warning if present
      if (cwdWarningPrefix) {
        warningParts.push(cwdWarningPrefix);
      }

      // Issue #2466: Add fork-session warning
      // Issue #2475: 自作Issueリストを表示
      if (forkSessionWarnings.length > 0) {
        const selfCreatedIssues = currentSession ? loadSessionCreatedIssues(currentSession) : [];
        let selfCreatedNote: string;
        if (selfCreatedIssues.length > 0) {
          const issueList = selfCreatedIssues.map((n) => `#${n}`).join(", ");
          selfCreatedNote = `\n**このセッションで作成したIssue**: ${issueList}\nこれらのIssueへの作業は**警告なしで許可**されています。\n`;
        } else {
          selfCreatedNote =
            "\n**このセッションで作成したIssue**: " +
            "（このセッション中に作成したIssueはまだありません）\n" +
            "今後このセッションで新しく作成したIssueへの作業は" +
            "**警告なしで許可**されます。\n";
        }

        warningParts.push(
          [
            "⚠️ **fork-session検出**: 元セッション（fork元）のworktreeがあります:",
            "",
            ...forkSessionWarnings,
            "",
            "**これらのworktreeへの介入は禁止です。**",
            "元セッションがまだ作業中の可能性があります。",
            selfCreatedNote,
            "その場合は、元セッションのworktreeとは**別の新しいworktree**を作成してください。",
            "",
          ].join("\n"),
        );
      }

      // Add worktree warnings if present
      if (warnings.length > 0) {
        warningParts.push(
          `以下のworktreeに注意が必要です:\n\n${warnings.join("\n")}\n\nこれらのworktreeに関連するIssueは作業中の可能性があります。\nAGENTS.mdのルールに従い、以下を確認してください:\n- worktreeがロック中 or 未コミット変更あり → 作業開始しない\n- 直近1時間以内のコミット → ユーザーに確認\n- 別セッションマーカー → 引き継がず他のIssueへ\n- 古いセッションマーカー → コンテキスト継続の可能性。ユーザーに確認\n`,
        );
      }

      let warningMsg = warningParts.slice(0, 2).join(" ");
      if (warningParts.length > 2) {
        warningMsg += `\n\n${warningParts.slice(2).join("\n\n")}`;
      }

      await logHookExecution(
        HOOK_NAME,
        "warn",
        `Found ${warnings.length} worktrees requiring attention, ` +
          `fork-session warnings: ${forkSessionWarnings.length}, ` +
          `CWD in worktree: ${Boolean(cwdWarningPrefix)}`,
      );
      result.systemMessage = warningMsg;
    } else {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `Checked ${worktreesInfo.length} worktrees, all clear`,
      );
    }
  } catch (error) {
    // Fail open - don't block on errors
    console.error(`[${HOOK_NAME}] Hook error: ${formatError(error)}`);
  }

  console.log(JSON.stringify(result));
}

// Only run main when executed directly, not when imported
if (import.meta.main) {
  main();
}
